README.md
# Simple Ceph RBD Docker VolumeDriver Plugin

* Use Case: Persistent Storage for a Single Docker Container
  * one RBD Image can only be used by one Docker Container at a time

* Plugin is a separate process running alongside Docker Daemon
  * plugin can be configured for a single Ceph User
  * run multiple plugin instances for varying configs (ceph user, default pool, default size)
  * OPTIONAL: pass extra config via volume name to override default pool and creation size:
    * docker run --volume-driver rbd -v poolname/imagename@size:/mnt/disk1 ...

* plugin supports all Docker VolumeDriver Plugin API commands (Volume Plugin API v1.12.x)
  * Create - can provision Ceph RBD Image in a pool of a certain size
    * controlled by `--create` boolean flag (default false)
    * default size from `--size` flag (default 20480 = 20GB)
  * Mount - Locks, Maps and Mounts RBD Image to the Host system
  * Unmount - Unmounts, Unmaps and Unlocks the RBD Image on request
  * Remove - Removes (destroys) RBD Image on request
    * only called for `docker run --rm -v ...` or `docker rm -v ...`
    * action controlled by plugin's `--remove` flag, which can be one of three values:
      - ''ignore'' - the call to delete the ceph rbd volume is ignored (default)
      - ''rename'' - will cause image to be renamed with _zz_ prefix for later culling
      - ''delete'' - will actually delete ceph rbd image (destructive)
  * Get, List - Return information on accessible RBD volumes

## Plugin Setup

Plugin is a standalone process and places a Socket file in a known location;
needs to start before Docker.  It does not daemonize by default, so if you need
it in the background, use normal shell process control (&).

The driver has a name, also used to name the socket, which is used to refer to
the plugin via the `--volume-driver=name` docker CLI option, allowing multiple
uniquely named plugin instances with different default configurations.

For the default name is "rbd", use `--volume-driver rbd` from docker.

General build/run requirements:
* /usr/bin/rbd for manipulating Ceph RBD images
* /usr/sbin/mkfs.xfs for fs creation (default fstype)
* /usr/bin/mount and /usr/bin/umount
* golang/dep tool

Tested with Ceph version 0.94.2 on Centos 7.1 host with Docker 1.12

### Building rbd-docker-plugin

Clone the repo and use the Makefile:

    make

To get `dist/rbd-docker-plugin` binary.

Or the equivalent shell commands:

    go get -u github.com/golang/dep/cmd/dep
    dep ensure
    go build -v -x -o dist/rbd-docker-plugin .

If none of the dependencies has changed (??) you might be able to get away with:

    go get github.com/porcupie/rbd-docker-plugin


### Commandline Options

    Usage of ./rbd-docker-plugin:
      --ceph-user="admin": Ceph user to use for RBD
      --create=false: Can auto Create RBD Images (default: false)
      --fs="xfs": FS type for the created RBD Image (must have mkfs.type)
      --logdir="/var/log": Logfile directory for RBD Docker Plugin
      --mount="/var/lib/docker/volumes": Mount directory for volumes on host
      --name="rbd": Docker plugin name for use on --volume-driver option
      --pool="rbd": Default Ceph Pool for RBD operations
      --remove=false: Can Remove (destroy) RBD Images (default: false, volume will be renamed zz_name)
      --size=20480: RBD Image size to Create (in MB) (default: 20480=20GB

### Start the Plugin

Start with the default options:

* socket name=rbd, pool=rbd, user=admin, logfile=/var/log/rbd-docker-plugin.log
* no creation or removal of volumes

    sudo rbd-docker-plugin

For Debugging: send log to STDERR:

    sudo RBD_DOCKER_PLUGIN_DEBUG=1 rbd-docker-plugin

Use a different socket name and Ceph pool

    sudo rbd-docker-plugin --name rbd2 --pool liverpool
    # docker run --volume-driver rbd2 -v ...

To allow creation of new RBD Images:

    sudo rbd-docker-plugin --create

To allow creation and removal:

    sudo rbd-docker-plugin --create --remove

Then you would be able to use RBD volumes via Docker CLI:

    docker run --volume-driver rbd -v ...

### Testing

Can test using docker engine 1.8+ which has `--volume-driver` support.

* https://docker.com/

Alternatively, you can POST json to the socket to manually test.  If your curl
is new enough (v7.40+), you can use the `--unix-socket` option and syntax.  You
can also use [this golang version](https://github.com/Soulou/curl-unix-socket)
instead:

    go get github.com/Soulou/curl-unix-socket


Once you have that you can POST json to the plugin:

    % sudo curl-unix-socket -v -X POST unix:///run/docker/plugins/rbd.sock:/Plugin.Activate
    > POST /Plugin.Activate HTTP/1.1
    > Socket: /run/docker/plugins/rbd.sock
    > Content-Length: 0
    >
    < HTTP/1.1 200 OK
    < Content-Type: appplication/vnd.docker.plugins.v1+json
    < Date: Tue, 28 Jul 2015 18:52:11 GMT
    < Content-Length: 33
    {"Implements": ["VolumeDriver"]}


    # Plugin started without --create:
    % sudo curl-unix-socket -v -X POST -d '{"Name": "testimage"}' unix:///run/docker/plugins/rbd.sock:/VolumeDriver.Create
    > POST /VolumeDriver.Create HTTP/1.1
    > Socket: /run/docker/plugins/rbd.sock
    > Content-Length: 21
    >
    < HTTP/1.1 500 Internal Server Error
    < Content-Length: 62
    < Content-Type: appplication/vnd.docker.plugins.v1+json
    < Date: Tue, 28 Jul 2015 18:53:20 GMT
    {"Mountpoint":"","Err":"Ceph RBD Image not found: testimage"}

    # Plugin started --create turned on will create unknown image:
    % sudo curl-unix-socket -v -X POST -d '{"Name": "testimage"}' unix:///run/docker/plugins/rbd.sock:/VolumeDriver.Create
    > POST /VolumeDriver.Create HTTP/1.1
    > Socket: /run/docker/plugins/rbd.sock
    > Content-Length: 21
    >
    < HTTP/1.1 200 OK
    < Content-Length: 27
    < Content-Type: appplication/vnd.docker.plugins.v1+json
    < Date: Fri, 14 Aug 2015 19:47:35 GMT
    {"Mountpoint":"","Err":""}

## Examples

If you need persistent storage for your application container, you can use a
Ceph Rados Block Device (RBD) as a persistent disk.

You can provision the Block Device and Filesystem first, or allow a
sufficiently configured Plugin instance create it for you.  This plugin can
create RBD images with XFS filesystem.

1. (Optional) Provision RBD Storage yourself
  * `sudo rbd create --size 1024 foo`
  * `sudo rbd map foo`  => /dev/rbd1
  * `sudo mkfs.xfs /dev/rbd1`
  * `sudo rbd unmap /dev/rbd1`
2. Or Run the RBD Docker Plugin with `--create` option flag and just request a volume
  * `sudo rbd-docker-plugin --create`
3. Requesting and Using Volumes
  * `docker run --volume-driver=rbd --volume foo:/mnt/foo -it ubuntu /bin/bash`
  * Volume "foo" will be locked, mapped and mounted to Host and bind-mounted to container at `/mnt/foo`
  * When container exits, the volume will be unmounted, unmapped and unlocked
  * You can control the RBD Pool and initial Size using this syntax sugar:
    * foo@1024 => pool=rbd (default), image=foo, size 1GB
    * deep/foo =>  pool=deep, image=foo and default `--size` (20GB)
    * deep/foo@1024 => pool=deep, image=foo, size 1GB
    - pool must already exist

### Misc

* Create RBD Snapshots: `sudo rbd snap create --image foo --snap foosnap`
* Resize RBD Volume:
  * set max size: `sudo rbd resize --size 2048 --image foo`
  * map/mount and then fix XFS: `sudo xfs_growfs -d /mnt/foo`


## Links


- [Legacy Plugins](https://docs.docker.com/engine/extend/legacy_plugins/)
  - [Volume plugins](https://docs.docker.com/engine/extend/plugins_volume/)

# Packaging

Using [tpkg](http://tpkg.github.io) to distribute and specify native package
dependencies.  Tested with Centos 7.1 and yum/rpm packages.


# License

This project is using the MIT License (MIT), see LICENSE file.

Copyright (c) 2015 YP LLC


CHANGELOG.md
# Change Log
All notable changes to project should be documented in this file.
We attempt to adhere to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added
### Removed
### Changed

## [2.0.1] - 2017-08-28
### Changed
- fixed List and Get methods to only return Mountpoint value if volume is
actually mounted.  Hopefully this matches the Docker Volume API better and
fixes some spurious volume creation (See #3, #6, #12, and
yp-engineering/rbd-docker-plugin#36)

## [2.0.0] - 2017-08-25
### Added
- add golang/dep support for a repeatable build

### Removed
- Removed all usage of go-ceph library.  Simplifies the code and now only
shelling out to 'rbd' executable

### Changed
- Updated docker/go-plugin-helpers and latest sdk api
- Fixed List() to call out to 'rbd ls' instead of relying on in memory per-host list

## [1.5.2] - 2017-08
### Changed
- Remove cron script installation from tpkg

## [1.5.1] - 2016-09-20
### Changed
- Update XFS checks to mount/unmount to clear any disk logs

## [1.5.0] - 2016-09-01
### Added
- Updated plugin to include Capabilities() method and new method
  signatures for Mount(MountRequest) and Unmount(UnmountRequest)
- add xfs filesystem check before mount - if xfs-repair -n returns
  error, Mount operation fails with note to manually repair

## [1.4.1] - 2016-08-05
### Changed
- Bug fix from bad upstream merge: provide openContext and
  shutdownContext

## [1.4.0] - 2016-08-05
### Added
- Merged upstream fixes for pull porcupie/rbd-docker-plugin#7

## [1.3.0] - 2016-08-05
### Added
- Added goroutines and timeouts to all shell commands, hoping to prevent
  propagation of hung external procs to docker daemon

## [1.2.2] - 2016-06-08
### Changed
- Docker is calling Unmount after failed Mount, causing trouble if
  device is still in use by another container (locked by same node).
The workaround / hack is to bail earlier in Unmount API call if rbd
unmap fails with busy device error. This can leave the device usable but
possibly in a funky state (unmounted from host but still mounted and
accessible inside container)
- related to porcupie/rbd-docker-plugin#5

## [1.2.1] - 2016-06-02
### Changed
- When rbd map does not return device name but no error, try default
  device path (/dev/rbd/<pool>/<image>). Issue porcupie/rbd-docker-plugin#4

## [1.2.0] - 2016-06-02
### Added
- Updated to pull in yp-engineering/rbd-docker-plugin v0.9.1.2, which
  includes support for Docker Volume Create options: size, pool, fstype

## [1.1.1] - 2016-04-15
### Changed
- Due to issue with golang 1.6 and strict Host header requirements, we
  cannot use golang 1.6 to compile our plugin since docker never sends
  correct Host header for plugin socket usage.
  * Recompiled with golang 1.5

## [1.1.0] - 2016-04-15
### Added
- added a cron job and check script for the Ceph configs and tpkg update

## [1.0.0] - 2016-04-15
### Changed
- bump major version with deprecated --remove boolean flag
- --remove flag now takes one of three values: ignore, delete or rename
  - ignore will just ignore the docker volume Remove call (new default)
  - delete will destroy the rbd volume on Remove request
  - rename will rename the rbd volume on Remove request, prefixed with `zz_`

## [0.5.0] - 2016-04-13
### Changed
- pulled latest from upstream yp-engineering/rbd-docker-plugin
  - add new docker volume api support (Get, List)
  - use ceph/go-ceph instead of noahdesu/go-ceph
  - use docker/go-plugins-helpers/ instead of calavera/dkvolume

## [0.4.2] - 2016-03-16
### Changed
- Update logrotate config to restart instead of reload
### Added 
- Some new marathon-tester configs for running a test container in
  Marathon/Mesos environment

## [0.4.1] - 2015-12-03
### Changed
- Update systemd service unit to add --config /etc/ceph/ceph.conf 
- Force default config file in main.go to /etc/ceph/ceph.conf

## [0.4.0] - 2015-12-03
### Changed
- Last ditch effort : Update all Plugin RBD functions to use CLI shell
  commands instead of go-ceph library
- Provide command line flag --go-ceph to use go-ceph lib, otherwise default
  now is shell CLI command via `rbd` binary

## [0.3.1] - 2015-11-30
### Changed
- Try to open RBD Image without read-only option (no effect)
- Try to use same client-id for every connection -- not possible in
  go-ceph
- Adding --conf options to external rbd operations (was having micro-osd
  issues)

## [0.3.0] - 2015-11-25
### Changed
- Update go-ceph import to use github.com/ceph/go-ceph instead of
  noahdesu/go-ceph
- Recreate ceph connection and pool context for every operation (don't 
  try to cache them)

## [0.2.2] - 2015-11-19
### Changed
- Disable the reload operation in systemd service unit, having issues
  with go-ceph lib and that operation (panics)
- Update the Image Rename and Remove functions to use go-ceph lib
  instead of shelling out to rbd binary
- Update the tpkg scripts to start the service on installation

## [0.2.1] - 2015-09-11
### Added
- Merged pull request with some RPM scripts for use in generic Redhat EL7 (Thanks Clement Laforet <sheepkiller@cotds.org>)

## [0.2.0] - 2015-08-25
### Changed
- Added micro-osd script for testing Ceph locally

## [0.1.9] - 2015-08-20
### Changed
- Added user ID and options to more shell rbd binary exec commands (Thanks Sébastien Han <seb@redhat.com>)
- Moving version definition from tpkg.yml to version.go
- Better blkid integration (Thanks Sébastien Han <seb@redhat.com>)


utils_test.go
package main

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestSh_success(t *testing.T) {
	out, err := sh("ls")
	assert.Nil(t, err, formatError("ls", err))
	assert.Contains(t, out, "driver_test.go")
}

func TestSh_fail(t *testing.T) {
	_, err := sh("false")
	assert.NotNil(t, err, formatError("false", err))
}

func TestShWithDefaultTimeout_triggerDefaultTimeout(t *testing.T) {
	// reset this global for the tests
	defaultShellTimeout = 2 * time.Second

	// sleep long enough to trigger timeout
	sleepSecs := "4"

	// use the default timeout - we want to trigger it
	_, err := shWithDefaultTimeout("sleep", sleepSecs)
	assert.NotNil(t, err, "Expected to get error for timeout")
	assert.Contains(t, err.Error(), "Reached TIMEOUT", "Expected 'Reached TIMEOUT' error")

	// reset
	defaultShellTimeout = 2 * 60 * time.Second
}

func TestShWithTimeout_timeoutZeroFail(t *testing.T) {
	// pass 0 as our duration to trigger the error
	_, err := shWithTimeout(0, "sleep", "1")
	assert.NotNil(t, err, "Expected to get error for duration")
	assert.Contains(t, err.Error(), "duration needs to be positive", "Expected duration validation error")
}

func TestShWithTimeout_cmdSucceeds(t *testing.T) {
	// make command sleep a bit
	sleepSecs := "1"

	// make timeout a bit longer so we don't trigger it
	timeout := 2 * time.Second

	// pass timeout and cmd shorter than that
	_, err := shWithTimeout(timeout, "sleep", sleepSecs)
	assert.Nil(t, err, "Expected success for command, not timeout")
}

func TestShWithTimeout_cmdTimesOut(t *testing.T) {
	// make timeout a bit shorter so we can trigger it
	timeout := 1 * time.Second

	// make command sleep a bit longer than that
	sleepSecs := "4"

	// pass our timeout and long sleep
	_, err := shWithTimeout(timeout, "sleep", sleepSecs)
	assert.NotNil(t, err, "Expected to get error for timeout")
	assert.Contains(t, err.Error(), "Reached TIMEOUT", "Expected 'Reached TIMEOUT' error")
}


driver_test.go
// Copyright 2015 YP LLC.
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file.
package main

// unit tests that don't rely on ceph

import (
	"flag"
	"fmt"
	"os"
	"testing"

	"github.com/docker/go-plugins-helpers/volume"
	"github.com/stretchr/testify/assert"
)

// TODO: tests that need ceph
// make fake cluster?
// use dockerized container with ceph for tests?

const (
	TEST_SOCKET_PATH             = "/tmp/rbd-test.sock"
	EXPECTED_ACTIVATION_RESPONSE = "{\"Implements\": [\"VolumeDriver\"]}"
)

var (
	testDriver cephRBDVolumeDriver
)

func TestMain(m *testing.M) {
	flag.Parse()
	cephConf := os.Getenv("CEPH_CONF")

	testDriver = newCephRBDVolumeDriver(
		"test",
		"",
		"admin",
		"rbd",
		volume.DefaultDockerRootDirectory,
		cephConf,
	)

	handler := volume.NewHandler(testDriver)
	// Serve won't return so spin off routine
	go handler.ServeUnix(TEST_SOCKET_PATH, 0)

	os.Exit(m.Run())
}

func TestLocalLockerCookie(t *testing.T) {
	assert.NotEqual(t, "HOST_UNKNOWN", testDriver.localLockerCookie())
}

func TestRbdImageExists_noName(t *testing.T) {
	f_bool, err := testDriver.rbdImageExists(testDriver.pool, "")
	assert.Equal(t, false, f_bool, fmt.Sprintf("%s", err))
}

func TestRbdImageExists_withName(t *testing.T) {
	t.Skip("This fails for many reasons. Need to figure out how to do this in a container.")
	err := testDriver.createRBDImage("rbd", "foo", 1, "xfs")
	assert.Nil(t, err, formatError("createRBDImage", err))
	t_bool, err := testDriver.rbdImageExists(testDriver.pool, "foo")
	assert.Equal(t, true, t_bool, formatError("rbdImageExists", err))
}

// cephRBDDriver.parseImagePoolNameSize(string) (string, string, int, error)
func TestParseImagePoolNameSize_name(t *testing.T) {
	pool, name, size := parseImageAndHandleError(t, "foo")

	assert.Equal(t, testDriver.pool, pool, "Pool should be same")
	assert.Equal(t, "foo", name, "Name should be same")
	assert.Equal(t, *defaultImageSizeMB, size, "Size should be same")
}

func TestParseImagePoolNameSize_complexName(t *testing.T) {
	pool, name, size := parseImageAndHandleError(t, "es-data1_v2.3")

	assert.Equal(t, testDriver.pool, pool, "Pool should be same")
	assert.Equal(t, "es-data1_v2.3", name, "Name should be same")
	assert.Equal(t, *defaultImageSizeMB, size, "Size should be same")
}

func TestParseImagePoolNameSize_withPool(t *testing.T) {
	pool, name, size := parseImageAndHandleError(t, "liverpool/foo")

	assert.Equal(t, "liverpool", pool, "Pool should be same")
	assert.Equal(t, "foo", name, "Name should be same")
	assert.Equal(t, *defaultImageSizeMB, size, "Size should be same")
}

func TestParseImagePoolNameSize_withSize(t *testing.T) {
	pool, name, size := parseImageAndHandleError(t, "liverpool/foo@1024")

	assert.Equal(t, "liverpool", pool, "Pool should be same")
	assert.Equal(t, "foo", name, "Name should be same")
	assert.Equal(t, 1024, size, "Size should be same")
}

func TestParseImagePoolNameSize_withPoolAndSize(t *testing.T) {
	pool, name, size := parseImageAndHandleError(t, "foo@1024")

	assert.Equal(t, testDriver.pool, pool, "Pool should be same")
	assert.Equal(t, "foo", name, "Name should be same")
	assert.Equal(t, 1024, size, "Size should be same")
}

// need a way to test the socket access using basic format - since this broke
// in golang 1.6 with strict Host header checking even if using Unix sockets.
// Requires socat and sudo
//
// Error response when built with golang 1.6: 400 Bad Request: missing required Host header
func TestSocketActivate(t *testing.T) {
	t.Skip("This test requires socket, which seems to need root privs to build. So this test fails if run as normal user. TODO: Find a proper workaround.")
	out, err := sh("bash", "-c", "echo \"POST /Plugin.Activate HTTP/1.1\r\n\" | sudo socat unix-connect:/tmp/rbd-test.sock STDIO")
	assert.Nil(t, err, formatError("socat plugin activate", err))
	assert.Contains(t, out, EXPECTED_ACTIVATION_RESPONSE, "Expecting Implements VolumeDriver message")

}

// Helpers
func formatError(name string, err error) string {
	return fmt.Sprintf("ERROR calling %s: %q", name, err)
}

func parseImageAndHandleError(t *testing.T, name string) (string, string, int) {
	pool, name, size, err := testDriver.parseImagePoolNameSize(name)
	assert.Nil(t, err, formatError("parseImagePoolNameSize", err))
	return pool, name, size
}


unlock_test.go
// Copyright 2015 YP LLC.
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file.
package main

// trying to write a small test to reproduce (un)locking issues

import (
	"bytes"
	"fmt"
	"log"
	"os/exec"
	"strings"
	"testing"

	//"github.com/stretchr/testify/assert"
	assert "github.com/stretchr/testify/require"
)

//"github.com/ceph/go-ceph/rbd"
//"github.com/ceph/go-ceph/rados"

var (
	//testDriver cephRBDVolumeDriver
	testPool  = "rbd"
	testImage = "rbd-test"
)

func TestShUnlockImage(t *testing.T) {
	// lock it first ... ?
	locker, err := testDriver.lockImage(testPool, testImage)
	if err != nil {
		log.Printf("WARN: Unable to lock image in preparation for test: %s", err)
		locker = testDriver.localLockerCookie()
	}

	// now unlock it
	err = testDriver.unlockImage(testPool, testImage, locker)
	assert.Nil(t, err, fmt.Sprintf("Unable to unlock image using sh rbd: %s", err))
}

func test_sh(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	log.Printf("DEBUG: sh CMD: %q", cmd)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	out, err := cmd.Output()
	if err != nil {
		log.Printf("ERROR: %q: %s", err, stderr)
	}
	return strings.Trim(string(out), " \n"), err
}


utils.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"log"
	"os/exec"
	"os/user"
	"strconv"
	"strings"
	"time"
)

var (
	defaultShellTimeout = 2 * 60 * time.Second
)

// returns current user gid or 0
func currentGid() int {
	gid := 0
	current, err := user.Current()
	if err != nil {
		return 0
	}
	gid, err = strconv.Atoi(current.Gid)
	if err != nil {
		return 0
	}

	return gid
}

// sh is a simple os.exec Command tool, returns trimmed string output
func sh(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	if isDebugEnabled() {
		log.Printf("DEBUG: sh CMD: %q", cmd)
	}
	// TODO: capture and output STDERR to logfile?
	out, err := cmd.Output()
	return strings.Trim(string(out), " \n"), err
}

// ShResult used for channel in timeout
type ShResult struct {
	Output string // STDOUT
	Err    error  // go error, not STDERR
}

type ShTimeoutError struct {
	timeout time.Duration
}

func (e ShTimeoutError) Error() string {
	return fmt.Sprintf("Reached TIMEOUT on shell command")
}

// shWithDefaultTimeout will use the defaultShellTimeout so you dont have to pass one
func shWithDefaultTimeout(name string, args ...string) (string, error) {
	return shWithTimeout(defaultShellTimeout, name, args...)
}

// shWithTimeout will run the Cmd and wait for the specified duration
func shWithTimeout(howLong time.Duration, name string, args ...string) (string, error) {
	// duration can't be zero
	if howLong <= 0 {
		return "", fmt.Errorf("Timeout duration needs to be positive")
	}
	// set up the results channel
	resultsChan := make(chan ShResult, 1)
	if isDebugEnabled() {
		log.Printf("DEBUG: shWithTimeout: %v, %s, %v", howLong, name, args)
	}

	// fire up the goroutine for the actual shell command
	go func() {
		out, err := sh(name, args...)
		resultsChan <- ShResult{Output: out, Err: err}
	}()

	select {
	case res := <-resultsChan:
		return res.Output, res.Err
	case <-time.After(howLong):
		return "", ShTimeoutError{timeout: howLong}
	}

	return "", nil
}

// grepLines pulls out lines that match a string (no regex ... yet)
func grepLines(data string, like string) []string {
	var result = []string{}
	if like == "" {
		log.Printf("ERROR: unable to look for empty pattern")
		return result
	}
	like_bytes := []byte(like)

	scanner := bufio.NewScanner(strings.NewReader(data))
	for scanner.Scan() {
		if bytes.Contains(scanner.Bytes(), like_bytes) {
			result = append(result, scanner.Text())
		}
	}
	if err := scanner.Err(); err != nil {
		log.Printf("WARN: error scanning string for %s: %s", like, err)
	}

	return result
}


micro-osd.sh
#
#    Copyright (C) 2013,2014 Loic Dachary <loic@dachary.org>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
set -e
set -u

DIR=$1

#if ! dpkg -l ceph ; then
# wget -q -O- 'https://ceph.com/git/?p=ceph.git;a=blob_plain;f=keys/release.asc' | sudo apt-key add -
# echo deb http://ceph.com/debian-dumpling/ $(lsb_release -sc) main | sudo tee /etc/apt/sources.list.d/ceph.list
# sudo apt-get update
# sudo apt-get --yes install ceph ceph-common
#fi

# get rid of process and directories leftovers
pkill ceph-mon || true
pkill ceph-osd || true
pkill ceph-mds || true
rm -fr $DIR

# cluster wide parameters
mkdir -p ${DIR}/log
cat >> $DIR/ceph.conf <<EOF
[global]
fsid = $(uuidgen)
osd crush chooseleaf type = 0
run dir = ${DIR}/run
auth cluster required = none
auth service required = none
auth client required = none
osd pool default size = 1
EOF
export CEPH_ARGS="--conf ${DIR}/ceph.conf"

# single monitor
MON_DATA=${DIR}/mon
mkdir -p $MON_DATA

cat >> $DIR/ceph.conf <<EOF
[mon.0]
log file = ${DIR}/log/mon.log
chdir = ""
mon cluster log file = ${DIR}/log/mon-cluster.log
mon data = ${MON_DATA}
mon addr = 127.0.0.1
EOF

ceph-mon --id 0 --mkfs --keyring /dev/null
touch ${MON_DATA}/keyring
ceph-mon --id 0 

# single osd
OSD_DATA=${DIR}/osd
mkdir ${OSD_DATA}

cat >> $DIR/ceph.conf <<EOF
[osd.0]
log file = ${DIR}/log/osd.log
chdir = ""
osd data = ${OSD_DATA}
osd journal = ${OSD_DATA}.journal
osd journal size = 100
osd objectstore = memstore
EOF

OSD_ID=$(ceph osd create)
ceph osd crush add osd.${OSD_ID} 1 root=default host=localhost
ceph-osd --id ${OSD_ID} --mkjournal --mkfs
ceph-osd --id ${OSD_ID}

# we don't need mds since we are not using cephfs - right?

# check that it works
rados --pool rbd put group /etc/group
rados --pool rbd get group ${DIR}/group
diff /etc/group ${DIR}/group
ceph osd tree

export CEPH_CONF="${DIR}/ceph.conf"


marathon-test/run.sh
#!/bin/sh

# append a hello to a log file

RBD_TEST=${RBD_TEST:-/mnt/foo}
LOG_FILE=${RBD_TEST}/hello.log


# don't make marathon churn too much ...
SLEEP_TIME=${SLEEP_TIME:-300}

echo "hello from $HOSTNAME"

# check for the file
LOG_ERROR=0
if [ ! -d $RBD_TEST ] ; then
    echo "ERROR: $HOSTNAME: unable to find rbd mount: $RBD_TEST"
    LOG_ERROR=1
fi

if [ ! -f $LOG_FILE ] ; then
    echo "ERROR: $HOSTNAME: unable to find log file: $LOG_FILE"
    LOG_ERROR=1
else
    echo "NOTE: found the existing mounted log file: $LOG_FILE ==>"
        cat $LOG_FILE
    echo "----"
fi

# append our note to log
echo "$HOSTNAME	`date`" | tee -a $LOG_FILE

# sleep a bit and exit
echo -n "sleeping $SLEEP_TIME ... "
sleep $SLEEP_TIME

echo "goodbye from $HOSTNAME"

if [ $LOG_ERROR != 0 ] ; then
    exit $LOG_ERROR
fi


main.go
// Copyright 2015 YP LLC.
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file.
package main

// Ceph RBD VolumeDriver Docker Plugin, setup config and go

import (
	"errors"
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"

	"github.com/docker/go-plugins-helpers/volume"
)

var (
	VALID_REMOVE_ACTIONS = []string{"ignore", "delete", "rename"}

	// Plugin Option Flags
	versionFlag        = flag.Bool("version", false, "Print version")
	debugFlag          = flag.Bool("debug", false, "Debug output")
	pluginName         = flag.String("name", "rbd", "Docker plugin name for use on --volume-driver option")
	cephUser           = flag.String("user", "admin", "Ceph user")
	cephConfigFile     = flag.String("config", "/etc/ceph/ceph.conf", "Ceph cluster config") // more likely to have config file pointing to cluster
	cephCluster        = flag.String("cluster", "", "Ceph cluster")                          // less likely to run multiple clusters on same hardware
	defaultCephPool    = flag.String("pool", "rbd", "Default Ceph Pool for RBD operations")
	pluginDir          = flag.String("plugins", "/run/docker/plugins", "Docker plugin directory for socket")
	rootMountDir       = flag.String("mount", volume.DefaultDockerRootDirectory, "Mount directory for volumes on host")
	logDir             = flag.String("logdir", "/var/log", "Logfile directory")
	canCreateVolumes   = flag.Bool("create", false, "Can auto Create RBD Images")
	defaultImageSizeMB = flag.Int("size", 20*1024, "RBD Image size to Create (in MB) (default: 20480=20GB)")
	defaultImageFSType = flag.String("fs", "xfs", "FS type for the created RBD Image (must have mkfs.type)")
)

// setup a validating flag for remove action
type removeAction string

func (a *removeAction) String() string {
	return string(*a)
}

func (a *removeAction) Set(value string) error {
	if !contains(VALID_REMOVE_ACTIONS, value) {
		return errors.New(fmt.Sprintf("Invalid value: %s, valid values are: %q", value, VALID_REMOVE_ACTIONS))
	}
	*a = removeAction(value)
	return nil
}

func contains(vals []string, check string) bool {
	for _, v := range vals {
		if check == v {
			return true
		}
	}
	return false
}

var removeActionFlag removeAction = "ignore"

func init() {
	flag.Var(&removeActionFlag, "remove", "Action to take on Remove: ignore, delete or rename")
	flag.Parse()
}

func socketPath() string {
	return filepath.Join(*pluginDir, *pluginName+".sock")
}

func logfilePath() string {
	return filepath.Join(*logDir, *pluginName+"-docker-plugin.log")
}

func main() {
	if *versionFlag {
		fmt.Printf("%s\n", VERSION)
		return
	}

	logFile, err := setupLogging()
	if err != nil {
		log.Fatalf("FATAL: Unable to setup logging: %s", err)
	}
	defer shutdownLogging(logFile)

	log.Printf("INFO: starting rbd-docker-plugin version %s", VERSION)
	log.Printf("INFO: canCreateVolumes=%v, removeAction=%q", *canCreateVolumes, removeActionFlag)
	log.Printf(
		"INFO: Setting up Ceph Driver for PluginID=%s, cluster=%s, ceph-user=%s, pool=%s, mount=%s, config=%s",
		*pluginName,
		*cephCluster,
		*cephUser,
		*defaultCephPool,
		*rootMountDir,
		*cephConfigFile,
	)

	// double check for config file - required especially for non-standard configs
	if *cephConfigFile == "" {
		log.Fatal("FATAL: Unable to use ceph rbd tool without config file")
	}
	if _, err = os.Stat(*cephConfigFile); os.IsNotExist(err) {
		log.Fatalf("FATAL: Unable to find ceph config needed for ceph rbd tool: %s", err)
	}

	// build driver struct -- but don't create connection yet
	d := newCephRBDVolumeDriver(
		*pluginName,
		*cephCluster,
		*cephUser,
		*defaultCephPool,
		*rootMountDir,
		*cephConfigFile,
	)

	log.Println("INFO: Creating Docker VolumeDriver Handler")
	h := volume.NewHandler(d)

	socket := socketPath()
	log.Printf("INFO: Opening Socket for Docker to connect: %s", socket)
	// ensure directory exists
	err = os.MkdirAll(filepath.Dir(socket), os.ModeDir)
	if err != nil {
		log.Fatalf("FATAL: Error creating socket directory: %s", err)
	}

	// setup signal handling after logging setup and creating driver, in order to signal the logfile and ceph connection
	// NOTE: systemd will send SIGTERM followed by SIGKILL after a timeout to stop a service daemon
	signalChannel := make(chan os.Signal, 2) // chan with buffer size 2
	signal.Notify(signalChannel, syscall.SIGTERM, syscall.SIGKILL)
	go func() {
		for sig := range signalChannel {
			//sig := <-signalChannel
			switch sig {
			case syscall.SIGTERM, syscall.SIGKILL:
				log.Printf("INFO: received TERM or KILL signal: %s", sig)
				shutdownLogging(logFile)
				os.Exit(0)
			}
		}
	}()

	// open socket
	err = h.ServeUnix(socket, currentGid())

	if err != nil {
		log.Printf("ERROR: Unable to create UNIX socket: %v", err)
	}
}

// isDebugEnabled checks for RBD_DOCKER_PLUGIN_DEBUG environment variable
func isDebugEnabled() bool {
	return *debugFlag || os.Getenv("RBD_DOCKER_PLUGIN_DEBUG") == "1"
}

// setupLogging attempts to log to a file, otherwise stderr
func setupLogging() (*os.File, error) {
	// use date, time and filename for log output
	log.SetFlags(log.LstdFlags | log.Lshortfile)

	// setup logfile - path is set from logfileDir and pluginName
	logfileName := logfilePath()
	if !isDebugEnabled() && logfileName != "" {
		logFile, err := os.OpenFile(logfileName, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
		if err != nil {
			// check if we can write to directory - otherwise just log to stderr?
			if os.IsPermission(err) {
				log.Printf("WARN: logging fallback to STDERR: %v", err)
			} else {
				// some other, more extreme system error
				return nil, err
			}
		} else {
			log.Printf("INFO: setting log file: %s", logfileName)
			log.SetOutput(logFile)
			return logFile, nil
		}
	}
	return nil, nil
}

func shutdownLogging(logFile *os.File) {
	// flush and close the file
	if logFile != nil {
		log.Println("INFO: closing log file")
		logFile.Sync()
		logFile.Close()
	}
}

func reloadLogging(logFile *os.File) (*os.File, error) {
	log.Println("INFO: reloading log")
	if logFile != nil {
		shutdownLogging(logFile)
	}
	return setupLogging()
}
